### DISCUSSION TITLE
Distributed Training

### DISCUSSION THREAD

**Discussion by jackmead515** [talmolab/sleap#1348] [Help!]
Hi! Quickly becoming a fan of this platform for it's ease of use and streamline API.

I was wondering if you could provide a short summary of what would be needed to do distributed training?

I have a fair amount of experience using Dask and have my own private cluster. I'd also be more than willing to write some code to support this.

**Comment by talmo** [comment]
Hi @jackmead515,

Glad you've been liking it! This has been a feature on the TODO list for years now -- we'd love a contribution!

Not sure how to use Dask to achieve this (maybe if it can parallelize `tf.data.Dataset` ops?), but the way you'd go about it would be to use [`tf.distribute.Strategy`](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy) to parallelize the training over multiple GPUs.

This tutorial shows the `tf.keras` side:
https://www.tensorflow.org/tutorials/distribute/keras

And this one the SLEAP side:
https://sleap.ai/notebooks/Interactive_and_resumable_training.html

You can probably start by scoping the whole `Trainer` class in a `tf.distribute.Strategy`. If you need to go low-level, this is probably a good place to start since it's where we instantiate the relevant TensorFlow objects:

https://github.com/talmolab/sleap/blob/1a0404c0ffae7b248eb360562b0bb95a42a287b6/sleap/nn/training.py#L721

Let us know if you have any questions and let us know how it goes if you get a chance to give it a spin!

Cheers,

Talmo

**Reply by jackmead515** [reply]
Heck yeah I'll give it a spin. Thanks for the links I'll polish up my knowledge in this frontier and offer a proposed solution.
